sklearn/neighbors/dist_metrics.pyx

```diff
@@ -1111,7 +1100,13 @@ cdef class PyFuncDistance(DistanceMetric):
         with gil:
             x1arr = _buffer_to_ndarray(x1, size)
             x2arr = _buffer_to_ndarray(x2, size)
-            return self.func(x1arr, x2arr, **self.kwargs)
+            d = self.func(x1arr, x2arr, **self.kwargs)
+            try:
+                return d
+            except TypeError:
```

Comment:
Hello @GaelVaroquaux 

I think test can be something like the following script,
it can show that this exception is indeed raised when customize distance function returns a non-float.
```python
from sklearn.neighbors import BallTree
import timeit
import numpy as np

def wrong_distance(x, y):
    return "1"

n_samples = 10 ** 3
n_dim = 10
X = np.asarray(range(n_samples * n_dim)).reshape(n_samples, n_dim)
b = BallTree(X, metric=wrong_distance)
```